File: tools/testing/connection/httpx/redis.py
#!/usr/bin/env python3
"""
Redis Connection Test (using redis-py)

Authentication: Password or ACL (username:password)
Protocol: Redis protocol (not HTTP)
Health Check: PING command

Note: Redis uses its own protocol, not HTTP. This file uses redis-py library.
For HTTP-based Redis (Redis REST API), see the httpx examples.
"""

import asyncio
import os
from typing import Any

# ============================================================================
# Configuration - Override these values
# ============================================================================

CONFIG = {
    # Required
    "REDIS_HOST": os.getenv("REDIS_HOST", "localhost"),
    "REDIS_PORT": int(os.getenv("REDIS_PORT", "6379")),
    "REDIS_PASSWORD": os.getenv("REDIS_PASSWORD", ""),
    "REDIS_USERNAME": os.getenv("REDIS_USERNAME", ""),  # For ACL (Redis 6+)
    "REDIS_DB": int(os.getenv("REDIS_DB", "0")),

    # Optional: TLS Configuration
    "REDIS_USE_SSL": os.getenv("REDIS_USE_SSL", "false").lower() == "true",
}


# ============================================================================
# Health Check (using redis-py)
# ============================================================================

async def health_check_redis_py() -> dict[str, Any]:
    """Perform health check using redis-py async."""
    print("\n=== Redis Health Check (redis-py) ===\n")

    try:
        import redis.asyncio as redis

        # Build connection URL
        auth = ""
        if CONFIG["REDIS_USERNAME"] and CONFIG["REDIS_PASSWORD"]:
            auth = f"{CONFIG['REDIS_USERNAME']}:{CONFIG['REDIS_PASSWORD']}@"
        elif CONFIG["REDIS_PASSWORD"]:
            auth = f":{CONFIG['REDIS_PASSWORD']}@"

        protocol = "rediss" if CONFIG["REDIS_USE_SSL"] else "redis"
        url = f"{protocol}://{auth}{CONFIG['REDIS_HOST']}:{CONFIG['REDIS_PORT']}/{CONFIG['REDIS_DB']}"

        print(f"Connecting to: {protocol}://{CONFIG['REDIS_HOST']}:{CONFIG['REDIS_PORT']}/{CONFIG['REDIS_DB']}")

        client = redis.from_url(url)

        # Test connection
        pong = await client.ping()
        print(f"PING: {pong}")

        # Get server info
        info = await client.info("server")
        print(f"Redis Version: {info.get('redis_version')}")
        print(f"OS: {info.get('os')}")

        await client.close()

        return {"success": True, "data": {"ping": pong, "version": info.get("redis_version")}}
    except ImportError:
        print("Error: redis package not installed. Run: pip install redis")
        return {"success": False, "error": "redis package not installed"}
    except Exception as e:
        print(f"Error: {e}")
        return {"success": False, "error": str(e)}


# ============================================================================
# Sample Operations
# ============================================================================

async def sample_operations() -> dict[str, Any]:
    """Perform sample Redis operations."""
    print("\n=== Sample Redis Operations ===\n")

    try:
        import redis.asyncio as redis

        # Build connection URL
        auth = ""
        if CONFIG["REDIS_USERNAME"] and CONFIG["REDIS_PASSWORD"]:
            auth = f"{CONFIG['REDIS_USERNAME']}:{CONFIG['REDIS_PASSWORD']}@"
        elif CONFIG["REDIS_PASSWORD"]:
            auth = f":{CONFIG['REDIS_PASSWORD']}@"

        protocol = "rediss" if CONFIG["REDIS_USE_SSL"] else "redis"
        url = f"{protocol}://{auth}{CONFIG['REDIS_HOST']}:{CONFIG['REDIS_PORT']}/{CONFIG['REDIS_DB']}"

        client = redis.from_url(url)

        # SET/GET
        await client.set("test:key", "hello world")
        value = await client.get("test:key")
        print(f"SET/GET: {value}")

        # HSET/HGET
        await client.hset("test:hash", mapping={"field1": "value1", "field2": "value2"})
        hash_value = await client.hgetall("test:hash")
        print(f"HSET/HGETALL: {hash_value}")

        # LIST
        await client.rpush("test:list", "item1", "item2", "item3")
        list_value = await client.lrange("test:list", 0, -1)
        print(f"RPUSH/LRANGE: {list_value}")

        # Cleanup
        await client.delete("test:key", "test:hash", "test:list")
        print("Cleanup: Deleted test keys")

        await client.close()

        return {"success": True}
    except ImportError:
        print("Error: redis package not installed. Run: pip install redis")
        return {"success": False, "error": "redis package not installed"}
    except Exception as e:
        print(f"Error: {e}")
        return {"success": False, "error": str(e)}


# ============================================================================
# HTTP-based Redis (Upstash REST API example)
# ============================================================================

async def health_check_upstash() -> dict[str, Any]:
    """Health check for Upstash Redis REST API."""
    print("\n=== Upstash Redis REST API Health Check ===\n")

    import httpx

    upstash_url = os.getenv("UPSTASH_REDIS_REST_URL", "")
    upstash_token = os.getenv("UPSTASH_REDIS_REST_TOKEN", "")

    if not upstash_url or not upstash_token:
        print("Upstash credentials not configured. Set UPSTASH_REDIS_REST_URL and UPSTASH_REDIS_REST_TOKEN")
        return {"success": False, "error": "Upstash credentials not configured"}

    async with httpx.AsyncClient() as client:
        try:
            response = await client.post(
                f"{upstash_url}/ping",
                headers={
                    "Authorization": f"Bearer {upstash_token}",
                },
            )

            data = response.json()
            print(f"Status: {response.status_code}")
            print(f"Response: {data}")

            return {"success": response.status_code == 200, "data": data}
        except Exception as e:
            print(f"Error: {e}")
            return {"success": False, "error": str(e)}


# ============================================================================
# Run Tests
# ============================================================================

async def main():
    """Run connection tests."""
    print("Redis Connection Test")
    print("=====================")
    print(f"Host: {CONFIG['REDIS_HOST']}:{CONFIG['REDIS_PORT']}")
    print(f"Database: {CONFIG['REDIS_DB']}")
    print(f"SSL: {CONFIG['REDIS_USE_SSL']}")

    await health_check_redis_py()
    # await sample_operations()
    # await health_check_upstash()  # For Upstash REST API


if __name__ == "__main__":
    asyncio.run(main())

File: tools/testing/connection/undici/redis.mjs
/**
 * Redis - Connection Test (using ioredis)
 *
 * Note: Redis uses its own protocol, not HTTP.
 * This file demonstrates connection testing with ioredis.
 * For HTTP-based Redis (like Redis REST API), use the httpRedis example.
 *
 * Authentication: Password or ACL (username:password)
 * Default Port: 6379
 * Health Command: PING
 *
 * TLS/SSL Options:
 *   NODE_TLS_REJECT_UNAUTHORIZED=0  - Ignore all certificate errors
 *   REQUEST_CA_BUNDLE=/path/to/ca   - Custom CA bundle file
 *   SSL_CERT_FILE=/path/to/cert     - Custom SSL certificate file
 *   NODE_EXTRA_CA_CERTS=/path/to/ca - Additional CA certificates
 */

import Redis from 'ioredis';
import { readFileSync } from 'node:fs';

// ============================================================================
// Configuration - Override these values
// ============================================================================

const CONFIG = {
  // Option 1: Full Connection URL
  REDIS_URL: process.env.REDIS_URL || '',

  // Option 2: Individual Components
  REDIS_HOST: process.env.REDIS_HOST || 'localhost',
  REDIS_PORT: parseInt(process.env.REDIS_PORT || '6379', 10),
  REDIS_USERNAME: process.env.REDIS_USERNAME || '',
  REDIS_PASSWORD: process.env.REDIS_PASSWORD || '',
  REDIS_DB: parseInt(process.env.REDIS_DB || '0', 10),
  REDIS_TLS: process.env.REDIS_TLS || 'false', // 'true' for TLS

  // Optional: TLS Configuration
  // Set NODE_TLS_REJECT_UNAUTHORIZED=0 to ignore certificate errors
  REJECT_UNAUTHORIZED: process.env.NODE_TLS_REJECT_UNAUTHORIZED !== '0',

  // Optional: Custom CA certificates
  // REQUEST_CA_BUNDLE or SSL_CERT_FILE - path to custom CA bundle
  // NODE_EXTRA_CA_CERTS - handled automatically by Node.js
  CA_BUNDLE: process.env.REQUEST_CA_BUNDLE || process.env.SSL_CERT_FILE || '',
};

// ============================================================================
// TLS Configuration Helper
// ============================================================================

function getTlsOptions() {
  const options = {
    rejectUnauthorized: CONFIG.REJECT_UNAUTHORIZED,
  };

  // Load custom CA bundle if specified
  if (CONFIG.CA_BUNDLE) {
    try {
      options.ca = readFileSync(CONFIG.CA_BUNDLE);
      console.log(`Using custom CA bundle: ${CONFIG.CA_BUNDLE}`);
    } catch (err) {
      console.warn(`Warning: Could not read CA bundle: ${err.message}`);
    }
  }

  return options;
}

// ============================================================================
// Create Redis Client
// ============================================================================

function createClient() {
  const tlsOptions = getTlsOptions();

  if (CONFIG.REDIS_URL) {
    console.log(`Using connection URL: ${CONFIG.REDIS_URL.replace(/:[^:@]+@/, ':***@')}`);
    return new Redis(CONFIG.REDIS_URL, {
      tls: CONFIG.REDIS_URL.startsWith('rediss://') ? tlsOptions : undefined,
    });
  }

  const useTls = CONFIG.REDIS_TLS === 'true' || CONFIG.REDIS_PORT === 25061;

  const options = {
    host: CONFIG.REDIS_HOST,
    port: CONFIG.REDIS_PORT,
    db: CONFIG.REDIS_DB,
  };

  if (CONFIG.REDIS_PASSWORD) {
    options.password = CONFIG.REDIS_PASSWORD;
  }

  if (CONFIG.REDIS_USERNAME) {
    options.username = CONFIG.REDIS_USERNAME;
  }

  if (useTls) {
    options.tls = tlsOptions;
  }

  console.log(`Connecting to: ${CONFIG.REDIS_HOST}:${CONFIG.REDIS_PORT}`);
  return new Redis(options);
}

// ============================================================================
// Health Check
// ============================================================================

async function healthCheck() {
  console.log('\n=== Redis Health Check ===\n');

  const client = createClient();

  try {
    const result = await client.ping();

    console.log(`PING Response: ${result}`);
    console.log('Connection successful!');

    return { success: result === 'PONG', result };
  } catch (error) {
    console.error('Error:', error.message);
    return { success: false, error: error.message };
  } finally {
    await client.quit();
  }
}

// ============================================================================
// Sample Commands
// ============================================================================

async function serverInfo() {
  console.log('\n=== Server Info ===\n');

  const client = createClient();

  try {
    const info = await client.info();

    // Parse key info
    const lines = info.split('\n');
    const version = lines.find((l) => l.startsWith('redis_version:'));
    const memory = lines.find((l) => l.startsWith('used_memory_human:'));
    const clients = lines.find((l) => l.startsWith('connected_clients:'));

    console.log(version);
    console.log(memory);
    console.log(clients);

    return { success: true, info };
  } catch (error) {
    console.error('Error:', error.message);
    return { success: false, error: error.message };
  } finally {
    await client.quit();
  }
}

async function basicOperations() {
  console.log('\n=== Basic Operations ===\n');

  const client = createClient();

  try {
    // SET
    const setResult = await client.set('test:key', 'Hello World');
    console.log(`SET test:key: ${setResult}`);

    // GET
    const getResult = await client.get('test:key');
    console.log(`GET test:key: ${getResult}`);

    // DEL
    const delResult = await client.del('test:key');
    console.log(`DEL test:key: ${delResult}`);

    // Verify deletion
    const verifyResult = await client.get('test:key');
    console.log(`GET test:key (after delete): ${verifyResult}`);

    return { success: true };
  } catch (error) {
    console.error('Error:', error.message);
    return { success: false, error: error.message };
  } finally {
    await client.quit();
  }
}

async function listKeys(pattern = '*') {
  console.log(`\n=== List Keys: ${pattern} ===\n`);

  const client = createClient();

  try {
    const keys = await client.keys(pattern);

    console.log(`Found ${keys.length} keys`);
    keys.slice(0, 20).forEach((key) => {
      console.log(`  - ${key}`);
    });

    return { success: true, keys };
  } catch (error) {
    console.error('Error:', error.message);
    return { success: false, error: error.message };
  } finally {
    await client.quit();
  }
}

async function getDatabaseSize() {
  console.log('\n=== Database Size ===\n');

  const client = createClient();

  try {
    const size = await client.dbsize();
    console.log(`Database size: ${size} keys`);

    return { success: true, size };
  } catch (error) {
    console.error('Error:', error.message);
    return { success: false, error: error.message };
  } finally {
    await client.quit();
  }
}

// ============================================================================
// Run Tests
// ============================================================================

async function main() {
  console.log('Redis Connection Test');
  console.log('=====================');
  console.log(`Host: ${CONFIG.REDIS_HOST}:${CONFIG.REDIS_PORT}`);
  console.log(`Database: ${CONFIG.REDIS_DB}`);
  console.log(`TLS: ${CONFIG.REDIS_TLS}`);
  console.log(`TLS Verify: ${CONFIG.REJECT_UNAUTHORIZED}`);
  console.log(`CA Bundle: ${CONFIG.CA_BUNDLE || 'System default'}`);

  await healthCheck();
  // await serverInfo();
  // await basicOperations();
  // await listKeys('*');
  // await getDatabaseSize();
}

main().catch(console.error);
